### PR TITLE
libbuiltin/kconfig: Modify coverage function dependencies

### PR DESCRIPTION
**Pull request by W-M-R**
*Note: Please adhere to [Contributing Guidelines](https://github.com/apache/nuttx/blob/master/CONTRIBUTING.md).*

## Summary
image
```
    gprof: Remove duplicate content
    1. arch/arm/src/cmake/gcc.cmake: The same judgment has been made in line 164
    2. boards/sim/sim/sim/scripts/Make.defs: arch/sim/src/Makefile also has in line 147
```
```
    gcov: Correct existing gcov configuration
    
    1. add CONFIG_COVERAGE_ALL to replace CONFIG_SCHED_GCOV_ALL
    2. Correct all SCHED_GCOV, SCHED_GCOV_ALL
```
`gcov: arm,arm64 add coverage global`
## Impact

*Update this section, where applicable, on how change affects users,
build process, hardware, documentation, security, compatibility, etc.*

## Testing

*Update this section with details on how did you verify the change,
what Host was used for build (OS, CPU, compiler, ..), what Target was
used for verification (arch, board:config, ..), etc. Providing build
and runtime logs from before and after change is highly appreciated.*

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



## Does this PR meet the NuttX Requirements? Please be concise

No, this PR does not fully meet the NuttX requirements as described. While it provides a summary of the changes, it lacks crucial information in the Impact and Testing sections.

**Missing Information:**

* **Impact:** The Impact section is essentially empty.  It needs to address *all* listed points, even if the answer is "NO".  Explicitly stating "NO" with a brief justification is better than leaving it blank. For example, "Impact on user: NO.  This change is purely internal and does not affect user applications." This PR specifically needs to address potential impact on debug capabilities and code size.
* **Testing:** The Testing section is also incomplete.  It provides no information about the build host or target used for testing.  It mentions "gcov: arm,arm64 add coverage global" but doesn't explain *how* this was tested or provide any logs demonstrating the change's effectiveness.  Sample logs from before and after the change are required to demonstrate that the intended behavior is achieved and no regressions are introduced.

**How to Improve this PR:**

1. **Expand the Summary:**  While the provided summary is a good start, it could be improved by explicitly stating the *reason* for removing the duplicate content and correcting the gcov configuration.  For example, "Remove duplicate content in cmake and Make.defs files to simplify the build process and avoid potential conflicts."  and "Correct gcov configuration to ensure proper code coverage analysis on arm and arm64 architectures using CONFIG_COVERAGE_ALL."
2. **Complete the Impact Section:** Address *all* points in the Impact section, even if the answer is "NO."  Explain the reasoning behind each answer.  For instance, if there is no user impact, state why. If the change affects code size (e.g., due to removing duplicate code), this should be mentioned.
3. **Thoroughly Document Testing:** Provide detailed information about the build host (OS, CPU architecture, compiler version) and the target platform (architecture, board, configuration).  Include actual log output from before and after the changes, demonstrating that the duplicate content is gone and that gcov functions correctly with the new configuration.  Show specific examples of gcov output showing the difference in coverage before and after.


By providing this missing information, the PR will be much stronger and easier for reviewers to evaluate, increasing the likelihood of it being accepted.

File: libs/libbuiltin/Kconfig [hunk 1]
```diff
@@ -10,6 +10,7 @@ config LIB_BUILTIN
 
 config LIB_COMPILER_RT
 	bool
+	depends on ARCH_TOOLCHAIN_CLANG
```
Comment (xiaoxiang781216) [inline libs/libbuiltin/Kconfig:13]: but libbuiltin should work with gcc too.

File: boards/sim/sim/sim/scripts/Make.defs [hunk 2]
```diff
@@ -78,10 +78,6 @@ ifeq ($(CONFIG_SCHED_GCOV_ALL),y)
   ARCHOPTIMIZATION += -fprofile-generate -ftest-coverage
 endif
 
-ifneq ($(CONFIG_PROFILE_ALL)$(CONFIG_SIM_PROFILE),)
-  ARCHOPTIMIZATION += -pg
```
Comment (xiaoxiang781216) [inline boards/sim/sim/sim/scripts/Make.defs:82]: could you point out the dup location? I can't find it.

File: arch/tricore/src/common/ToolchainGnuc.defs [hunk 3]
```diff
@@ -58,8 +58,12 @@ ifneq ($(CONFIG_STACK_USAGE_WARNING),0)
   ARCHOPTIMIZATION += -Wstack-usage=$(CONFIG_STACK_USAGE_WARNING)
 endif
 
-ifeq ($(CONFIG_SCHED_GCOV_ALL),y)
-  ARCHOPTIMIZATION += -fprofile-generate -ftest-coverage
+ifeq ($(CONFIG_COVERAGE_ALL),y)
+  ifeq ($(CONFIG_ARCH_TOOLCHAIN_GCC),y)
+    ARCHOPTIMIZATION += -fprofile-generate -ftest-coverage
+  else ifeq ($(CONFIG_ARCH_TOOLCHAIN_CLANG),y)
```
Comment (xiaoxiang781216) [inline arch/tricore/src/common/ToolchainGnuc.defs:64]: why need for tricore?

File: arch/arm/src/cmake/armclang.cmake [hunk 4]
```diff
@@ -116,8 +116,8 @@ if(CONFIG_STACK_USAGE_WARNING AND NOT "${CONFIG_STACK_USAGE_WARNING}" STREQUAL
   add_compile_options(-Wstack-usage=${CONFIG_STACK_USAGE_WARNING})
 endif()
 
-if(CONFIG_SCHED_GCOV)
-  add_compile_options(-fprofile-generate -ftest-coverage)
+if(CONFIG_COVERAGE_ALL)
+  add_compile_options(-fprofile-instr-generate -fcoverage-mapping)
```
Comment (xiaoxiang781216) [inline arch/arm/src/cmake/armclang.cmake:120]: do we need different option for full and mini

File: arch/arm/src/cmake/clang.cmake [hunk 5]
```diff
@@ -117,8 +117,8 @@ if(CONFIG_STACK_USAGE_WARNING AND NOT "${CONFIG_STACK_USAGE_WARNING}" STREQUAL
   add_compile_options(-Wstack-usage=${CONFIG_STACK_USAGE_WARNING})
 endif()
 
-if(CONFIG_SCHED_GCOV_ALL)
-  add_compile_options(-fprofile-generate -ftest-coverage)
+if(CONFIG_COVERAGE_ALL)
+  add_compile_options(-fprofile-instr-generate -fcoverage-mapping)
```
Comment (xiaoxiang781216) [inline arch/arm/src/cmake/clang.cmake:121]: ditto

File: arch/arm/src/cmake/platform.cmake [hunk 6]
```diff
@@ -83,6 +83,6 @@ if(CONFIG_LIBSUPCXX_TOOLCHAIN)
   nuttx_find_toolchain_lib(libsupc++.a)
 endif()
 
-if(CONFIG_SCHED_GCOV)
+if(CONFIG_COVERAGE_TOOLCHAIN)
   nuttx_find_toolchain_lib(libgcov.a)
```
Comment (xiaoxiang781216) [inline arch/arm/src/cmake/platform.cmake:87]: do you need different library for clang

File: arch/arm/src/common/Toolchain.defs [hunk 7]
```diff
@@ -502,7 +510,7 @@ ifeq ($(CONFIG_LIBSUPCXX_TOOLCHAIN),y)
   EXTRA_LIBS += $(wildcard $(shell $(CC) $(ARCHCPUFLAGS) --print-file-name=libsupc++.a))
 endif
 
-ifeq ($(CONFIG_SCHED_GCOV),y)
+ifeq ($(CONFIG_COVERAGE_TOOLCHAIN),y)
   EXTRA_LIBS += $(wildcard $(shell $(CC) $(ARCHCPUFLAGS) --print-file-name=libgcov.a))
```
Comment (xiaoxiang781216) [inline arch/arm/src/common/Toolchain.defs:514]: ditto

File: arch/arm64/src/Toolchain.defs [hunk 8]
```diff
@@ -215,7 +223,7 @@ ifeq ($(CONFIG_LIBSUPCXX_TOOLCHAIN),y)
   EXTRA_LIBS += $(wildcard $(shell $(CC) $(ARCHCPUFLAGS) --print-file-name=libsupc++.a))
 endif
 
-ifeq ($(CONFIG_SCHED_GCOV),y)
+ifeq ($(CONFIG_COVERAGE_TOOLCHAIN),y)
   EXTRA_LIBS += $(wildcard $(shell $(CC) $(ARCHCPUFLAGS) --print-file-name=libgcov.a))
```
Comment (xiaoxiang781216) [inline arch/arm64/src/Toolchain.defs:227]: ditto

File: boards/sim/sim/sim/scripts/Make.defs (hunk 2 above)
Comment (W-M-R) [reply boards/sim/sim/sim/scripts/Make.defs:82]: arch/sim/src/Makefile:147 There is also this configuration

File: libs/libbuiltin/Kconfig (hunk 1 above)
Comment (W-M-R) [reply libs/libbuiltin/Kconfig:13]: This is the error reported after switching to the gcc compiler. Does this error need to be fixed or does it mean that compilation is not supported?
```
AS:  compiler-rt/compiler-rt/lib/builtins/arm/clzdi2.S compiler-rt/compiler-rt/lib/builtins/arm/chkstk.S: Assembler messages:
compiler-rt/compiler-rt/lib/builtins/arm/chkstk.S:28: Error: lo register required -- `sub r12,r12,#4096'
compiler-rt/compiler-rt/lib/builtins/arm/chkstk.S:29: Error: instruction not supported in Thumb16 mode -- `subs r5,r5,#4096'
compiler-rt/compiler-rt/lib/builtins/arm/chkstk.S:30: Error: lo register required -- `ldr r6,[r12]'
make[2]: *** [Makefile:45：bin/compiler-rt/compiler-rt/lib/builtins/arm/chkstk.o] 错误 1
```

File: arch/arm/src/cmake/armclang.cmake (hunk 4 above)
Comment (W-M-R) [reply arch/arm/src/cmake/armclang.cmake:120]: I think we don't need to do this for now, we can keep the same options as mini.profile. Other options are slightly different, we can explore them when we need them

File: arch/arm/src/cmake/platform.cmake (hunk 6 above)
Comment (W-M-R) [reply arch/arm/src/cmake/platform.cmake:87]: This CONFIG_COVERAGE_TOOLCHAIN ​​depends on gcc. When using clang, we don't need to link

File: boards/sim/sim/sim/scripts/Make.defs (hunk 2 above)
Comment (xiaoxiang781216) [reply boards/sim/sim/sim/scripts/Make.defs:82]: should we remove the dup from arch/sim/src/Makefile?

Comment (W-M-R) [reply boards/sim/sim/sim/scripts/Make.defs:82]: OK

File: libs/libbuiltin/Kconfig (hunk 1 above)
Comment (xiaoxiang781216) [reply libs/libbuiltin/Kconfig:13]: let's fix it.

File: libs/libbuiltin/Kconfig [hunk 9]
```diff
@@ -19,6 +19,7 @@ choice
 
 config BUILTIN_COMPILER_RT
 	bool "Builtin LLVM libclang_rt.builtins"
+	depends on ARCH_TOOLCHAIN_CLANG
```
Comment (xiaoxiang781216) [inline libs/libbuiltin/Kconfig:22]: ARCH_TOOLCHAIN_GNU

Comment (xiaoxiang781216) [reply libs/libbuiltin/Kconfig:22]: add depends on ARCH_TOOLCHAIN_CLANG at line 40

File: libs/libbuiltin/compiler-rt/Make.defs [hunk 10]
```diff
@@ -109,6 +109,8 @@ else ifeq ($(CONFIG_COVERAGE_MINI),y)
 
 CSRCS += coverage.c
 
+coverage.o: CFLAGS += -fno-profile-instr-generate -fno-coverage-mapping
```
Comment (xiaoxiang781216) [inline libs/libbuiltin/compiler-rt/Make.defs:112]: what's about InstrProfilingPlatform.c

**Pull request merged**